### PR TITLE
Misc audio fixes

### LINKED ISSUE #796: Implement downmixing on SoundIO backend
We need to handle channel downmixing on the SoundIO backend.
This is required for games like Pokémon Let's Go that uses 6 channels.

**NOTE:** As OpenAL backend handles dowmixing for us, it might be a good idea to detect for OpenAL first and fallback to SoundIO if nothing is available (as SoundIO is always available in our releases).

## Required by:
* https://github.com/Ryujinx/Ryujinx-Games-List/issues/115

### PR DESCRIPTION
**Pull request by Thog**
Changes:
- Implement software surround downmixing (fix #796).
- Fix a crash when no audio renderer were created when stopping emulation
- ~~The audio renderer now support output to a 5.1 surround compatible device.~~ => need more testing, removed.

NOTE: This PR also disable support of 5.1 surround on the OpenAL backend as we cannot detect if the hardware directly support it. (the downmixing applied by OpenAL on Windows is terribly slow)

### CODE REVIEW

File: Ryujinx.Audio/Renderers/OpenAL/OpenALAudioOut.cs [hunk 1]
```diff
@@ -104,6 +104,18 @@ private void AudioPollerWork()
             _context.Dispose();
         }
 
+        public bool SupportsChannelCount(int channels)
+        {
+            if (channels == 6)
+            {
+                // NOTE: OpenAL doesn't give us a way to know if the 5.1 setup is supported by hardware or actually emulated.
+                // TODO: find a way to determine hardware support.
+                // return AL.GetEnumValue("AL_FORMAT_51CHN16") != 0;
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/OpenAL/OpenALAudioOut.cs:113]: I think it's better to remove this code, as it doesn't work anyway, and only leave the comment.

**Comment by Thog** [comment]
Renamed and ready for review again

**Comment by Thog** [comment]
Implemented stereo to mono downmixing and fast path for downmixing. (using FMA and SSE2)

File: Ryujinx.Audio/Downmixing.cs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+                            Vector128<float> result = Vector128.Create(0.5f);
+
+                            result = Sse.Add(result, Sse.Multiply(back, backCoefficient));
+                            result = Sse.Add(result, Sse.Multiply(lowFrequency, lowFrequencyCoefficient));
+                            result = Sse.Add(result, Sse.Multiply(center, centerCoefficient));
+                            result = Sse.Add(result, Sse.Multiply(front, frontCoefficient));
+
+                            Sse2.Store(ptr + i * 2,
+                                       Sse2.PackSignedSaturate(Sse2.ConvertToVector128Int32(result), Vector128<int>.Zero));
```
Comment (gdkchan) [inline Ryujinx.Audio/Downmixing.cs:206]: nit: That could be in a single line (also for consistency with the function below).

File: Ryujinx.Audio/IAalOutput.cs [hunk 3]
```diff
@@ -4,15 +4,42 @@ namespace Ryujinx.Audio
...
+            }
+
+            switch (targetChannelCount)
+            {
+                case 6:
+                    return SelectHardwareChannelCount(2);
+                case 2:
+                    return SelectHardwareChannelCount(1);
+                case 1:
+                    throw new InvalidOperationException($"No valid channel configuration found!");
```
Comment (gdkchan) [inline Ryujinx.Audio/IAalOutput.cs:23]: I think this should be `ArgumentException`,  as the argument is invalid.

File: Ryujinx.Audio/IAalOutput.cs [hunk 4]
```diff
@@ -4,15 +4,42 @@ namespace Ryujinx.Audio
...
+                default:
+                    throw new InvalidOperationException($"Invalid targetChannelCount {targetChannelCount}");
```
Comment (gdkchan) [inline Ryujinx.Audio/IAalOutput.cs:25]: Same comment as above.

File: Ryujinx.Audio/Renderers/OpenAL/OpenALAudioOut.cs [hunk 5]
```diff
@@ -192,21 +201,47 @@ public long[] GetReleasedBuffers(int trackId, int maxCount)
...
+
+                        if (track.VirtualChannels == 6)
+                        {
+                            downmixedBuffer = Downmixing.DownMixSurroundToStereo(buffer);
+
+                            if (track.HardwareChannels == 1)
+                            {
+                                downmixedBuffer = Downmixing.DownMixStereoToMono(downmixedBuffer);
+                            }
+
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/OpenAL/OpenALAudioOut.cs:228]: nit: Empty line not necessary.

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs [hunk 6]
```diff
@@ -490,24 +504,62 @@ public void Stop()
...
 
-            // Calculate the size of the audio samples
-            int size = Unsafe.SizeOf<T>();
+            int sampleSize = Unsafe.SizeOf<T>();
+            int targetSize = sampleSize * buffer.Length;
+
+            // Do we need to downmix?
+            if (_hardwareChannels != _virtualChannels)
+            {
+                if (sampleSize != Unsafe.SizeOf<short>())
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs:520]: You can just use `sizeof(short)` instead of `Unsafe.SizeOf<short>()`.

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs [hunk 7]
```diff
@@ -490,24 +504,62 @@ public void Stop()
         /// <typeparam name="T">The audio sample type</typeparam>
         /// <param name="bufferTag">The unqiue tag of the buffer being appended</param>
         /// <param name="buffer">The buffer to append</param>
-        public void AppendBuffer<T>(long bufferTag, T[] buffer)
+        public void AppendBuffer<T>(long bufferTag, ReadOnlySpan<T> buffer) where T: struct
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs:507]: I think the contraint here should be `where T : unmanaged`.

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoRingBuffer.cs [hunk 8]
```diff
@@ -107,16 +109,18 @@ private void SetCapacity(int capacity)
         /// <summary>
         /// Writes a sequence of bytes to the ring buffer
         /// </summary>
-        /// <param name="buffer">A byte array containing the data to write</param>
+        /// <param name="buffer">A span containing the data to write</param>
         /// <param name="index">The zero-based byte offset in <paramref name="buffer" /> from which to begin copying bytes to the ring buffer</param>
-        /// <param name="count">The number of bytes to write</param>
-        public void Write<T>(T[] buffer, int index, int count)
+        /// <param name="count">The number of element to write</param>
+        public void Write<T>(ReadOnlySpan<T> buffer, int index, int count) where T: struct
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/SoundIo/SoundIoRingBuffer.cs:115]: `unmanaged`.

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs [hunk 9]
```diff
@@ -490,24 +504,62 @@ public void Stop()
...
-            int size = Unsafe.SizeOf<T>();
+            int sampleSize = Unsafe.SizeOf<T>();
+            int targetSize = sampleSize * buffer.Length;
+
+            // Do we need to downmix?
+            if (_hardwareChannels != _virtualChannels)
+            {
+                if (sampleSize != Unsafe.SizeOf<short>())
+                {
+                    throw new NotImplementedException("Non PCM16 downmixing isn't supported!");
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs:522]: I'd reword the exception message. Maybe `Only PCM16 downmixing is support` or `Downmixing formats other than PCM16 is not supported` would be more clear.

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs (hunk 6 above)
Comment (gdkchan) [reply Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs:520]: According to MSDN, looks like you can even do `sizeof(T)` if you put in the `unmanaged` constraint.

File: Ryujinx.Audio/IAalOutput.cs [hunk 10]
```diff
@@ -20,9 +20,9 @@ private int SelectHardwareChannelCount(int targetChannelCount)
                 case 2:
                     return SelectHardwareChannelCount(1);
                 case 1:
-                    throw new InvalidOperationException($"No valid channel configuration found!");
+                    throw new ArgumentException($"No valid channel configuration found!");
```
Comment (gdkchan) [inline Ryujinx.Audio/IAalOutput.cs:23]: No need for the `$` here (sorry, missed on the first review).

File: Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs [hunk 11]
```diff
@@ -504,22 +504,22 @@ public void Stop()
...
-        public void AppendBuffer<T>(long bufferTag, ReadOnlySpan<T> buffer) where T: struct
+        public void AppendBuffer<T>(long bufferTag, ReadOnlySpan<T> buffer) where T: unmanaged
         {
             if (AudioStream == null)
             {
                 return;
             }
 
-            int sampleSize = Unsafe.SizeOf<T>();
+            int sampleSize = sizeof(short);
```
Comment (gdkchan) [inline Ryujinx.Audio/Renderers/SoundIo/SoundIoAudioTrack.cs:514]: `sizeof(T)`

File: Ryujinx.Audio/Downmixing.cs [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+                downmixedBuffer[i * 2 + 1] = DownMixSurroundToStereo(coefficients, channel.BackRight, channel.LowFrequency, channel.FrontCenter, channel.FrontRight);
+            }
+
+            return downmixedBuffer;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static short[] DownMixStereoToMono(ReadOnlySpan<int> coefficients, ReadOnlySpan<short> data)
+        {
+            const int SurroundChannelCount = 2;
```
Comment (riperiperi) [inline Ryujinx.Audio/Downmixing.cs:226]: These variable names are incorrect.

Comment (Thog) [reply Ryujinx.Audio/Downmixing.cs:226]: oops
